core/source/Time.mint

```diff
@@ -21,185 +36,617 @@ module Time {
+
+    Time.dayOfWeekNumber(Time.utcDate(2018, 4, 5)) == 4
+  */
+  fun dayOfWeekNumber (time : Time) : Number {
     `
     (() => {
-      switch (#{what}) {
-        case 'month':
-          return DateFNS.endOfMonth(#{date})
-        case 'week':
-          return DateFNS.endOfWeek(#{date}, { weekStartsOn: 1 })
-        case 'day':
-          return DateFNS.endOfDay(#{date})
-        default:
-          return #{date}
-      }
+      const _ = #{time}.getUTCDay()
+      return _ === 0 ? 7 : _;
     })()
     `
   }
 
-  /* Returns an array of days from the given start to given end time. */
-  fun range (from : Time, to : Time) : Array(Time) {
+  /*
+  Returns the day of week of the given time.
+
+    Time.dayOfWeek(Time.utcDate(2018, 4, 5)) == Weekday::Thursday
+  */
+  fun dayOfWeek (time : Time) : Weekday {
+    case (dayOfWeekNumber(time)) {
+      1 => Weekday::Monday
+      2 => Weekday::Tuesday
+      3 => Weekday::Wednesday
+      4 => Weekday::Thursday
+      5 => Weekday::Friday
+      6 => Weekday::Saturday
+      => Weekday::Sunday
+    }
+  }
+
+  /*
+  Returns the day of month of the given time.
+
+    Time.dayOfMonth(Time.utcDate(2018, 4, 5)) == 5
+  */
+  fun dayOfMonth (time : Time) : Number {
+    `#{time}.getUTCDate()`
+  }
+
+  /*
+  Returns the day of the year of the given time.
+
+    Time.dayOfYear(Time.utcDate(2018, 4, 5)) == 95
+  */
+  fun dayOfYear (time : Time) : Number {
+    `
+    (() => {
+      const first =
+        Date.UTC(#{time}.getUTCFullYear(), 0, 1)
+
+      const current =
+        Date.UTC(#{time}.getUTCFullYear(), #{time}.getUTCMonth(), #{time}.getUTCDate())
+
+      return ((current - first) / 24 / 60 / 60 / 1000) + 1
+    })()
+    `
+  }
+
+  /*
+  Returns the hour of the given time.
+
+    Time.hour(Time.utc(2018, 4, 5, 10, 25, 30, 40) == 10
+  */
+  fun hour (time : Time) : Number {
+    `#{time}.getUTCHours()`
+  }
+
+  /*
+  Returns the minute of the given time.
+
+    Time.minute(Time.utc(2018, 4, 5, 10, 25, 30, 40)) == 25
+  */
+  fun minute (time : Time) : Number {
+    `#{time}.getUTCMinutes()`
+  }
+
+  /*
+  Returns the second of the given time.
+
+    Time.second(Time.utc(2018, 4, 5, 10, 25, 30, 40)) == 30
+  */
+  fun second (time : Time) : Number {
+    `#{time}.getUTCSeconds()`
+  }
+
+  /*
+  Returns the millisecond of the given time.
+
+    Time.millisecond(Time.utc(2018, 4, 5, 10, 25, 30, 40)) == 40
+  */
+  fun millisecond (time : Time) : Number {
+    `#{time}.getUTCMilliseconds()`
+  }
+
+  /*
+  Returns if the year of the given time is a leap year or not.
+
+    Time.isLeapYear(Time.utcDate(2011,1,1)) == false
+    Time.isLeapYear(Time.utcDate(2012,1,1)) == true
+  */
+  fun isLeapYear (time : Time) : Bool {
+    Number.isLeapYear(year(time))
+  }
+
+  /* MANIPULATION ------------------------------------------------------------ */
+
+  /*
+  Shifts the given time using the given time span.
+
+    Time.shift(Time.utcDate(2018, 4, 5), Time.Span::Days(2)) ==
```

Comment:
```suggestion
    (Time.utcDate(2018, 4, 5) |> Time.shift(Time.Span::Days(2))) ==
```

Reply:
Both these look equally nice to me, but neither looks near as nice as Crystal's `+ 2.days`, wish we could capture a greater economy of expression here.
Since `Time.shift` can only accept a `Time.Span::T(n)`, could we perhaps allow the compiler to accept a plain `T(n)` which is automatically cast as a `Time.Span`? That would at least give us
```
Time.utcDate(2018, 4, 5) |> Time.shift(Days(2))
```

Then picture another pipe syntax `|.` which automatically keeps us in the same module (`Time` in this case) as the last expression. That gets us down to:

```
Time.utcDate(2018, 4, 5) |. shift(Days(2))
```

Then we're getting somewhere.

Reply:
> Both these look equally nice to me, but neither looks near as nice as Crystal's `+ 2.days`, wish we could capture a greater economy of expression here. Since `Time.shift` can only accept a `Time.Span::T(n)`, could we perhaps allow the compiler to accept a plain `T(n)` which is automatically cast as a `Time.Span`? That would at least give us
> 
> ```
> Time.utcDate(2018, 4, 5) |> Time.shift(Days(2))
> ```
> 
> Then picture another pipe syntax `|.` which automatically keeps us in the same module (`Time` in this case) as the last expression. That gets us down to:
> 
> ```
> Time.utcDate(2018, 4, 5) |. shift(Days(2))
> ```
> 
> Then we're getting somewhere.

`Days(2)` looks nice at first but it would require leaving a bunch of classes unscoped, i.e., `Hours`, `Minutes`, `Seconds`, `Days`, `Weeks`, `Months` and `Years`. Time.days(2) seems better to me, and internally it could make the conversion, i.e.:
```
module TIme {
  fun days(amount : Number) : Time.Span::Days {
    Time.Span::Days(amount)
  }
}
```

That would also solve the problem with Weeks, as `Time.weeks` could use days in the background, i.e.:
```
module TIme {
  fun weeks(amount : Number) : Time.Span::Days {
    Time.Span::Days(amount * 7)
  }
}
```

Reply:
It would still end up as `|> Time.shift(Time.days(2))` - repeating `Time` over and over. I have this pattern frequently in my code where I repeat the same module name ad nauseum, so I think it would be appropriate to add a language-level feature for brevity.

Reply:
> It would still end up as `Time.utcDate(2018, 4, 5) |> Time.shift(Time.days(2))` - repeating `Time` over and over. I have this pattern frequently in my code where I repeat the same module name ad nauseum, so I think it would be appropriate to add a language-level feature for brevity.
> 
> Compare to Crystal's `Time.utc(2018, 4, 5) + 2.days` - no ambiguity there and no repetition. I'm always softly dismayed when I see something that lends itself to simple, clear representation in Crystal but it ends up boilerplatey and obfuscated in Mint.

It's hard to find examples that adequately portray what we're thinking, and we have to decide how much time we're willing to invest in picking/perfecting examples and when it's sufficient, but luckily I'm mostly a Ruby programmer so it's ridiculously easy to understand your point, which is all about the benefits of having a chaining/object-oriented approach.

In Rails I can simply write `Time.utc(2018,4,3) + 2.days` and it'll be equal to April 5th, 2018, UTC. Mint favors the functional approach, i.e., return values should be passed to functions, not be used to call functions using `.something`. There are pros and cons for both of those approaches, but it is what it is, Crystal is a good example of the absolute opposite approach, I've actually participated in a discussion about implementing pipes in Crystal (Like the `|>` operator in Mint) and there's a lot of resistance against it, because Crystal is all about OO and method chaining. A spin-off of Mint that had those would be something VERY interesting, but I, as many, currently lean towards  the Functional approach, which I like quite a lot, but it does, indeed, have a few issues. And Boilerplate is indeed one of the biggest problems, I believe. I avoided going into more detail to keep it short and focus on the `Time` module which is the topic here, but this is so important it deserves some clarification, so let's get to it:

Scoping is very important, this isn't obvious when talking about `Days` but if we consider a slightly larger system, let's take a simple blog/website as an example, it gets icky. Because imagine you start with `Post`s and `Page`s, plus the basic stuff, then you add `Thumbnail`s and `Category`(ies) for `Post`s, and `User`s. Then you decide on having `Thumbnail`s for `User`s, the name is taken already but it's ok, you just use `Avatar`, but then your boss tells you they want to add thumbnails for the categories, and you realize there's no sane way of making it work without having any ambiguity and with everything being easily recognizable and traceable and obvious. If you used `Post.Thumbnail`, `User.Thumbnail` and `Post.Category` it's actually not a problem, you just add `Post.Category.Thumbnail` and it never even poses any problem at all. That's the magic of scoping stuff, and it's important, and I actually believe we shouldn't mess with that.

Still, there is a nice option that would make both of us happy with the result, which is allowing access to scoped stuff withoug specifying the whole scope. Elm does that magnificently, and I miss how easy Elm makes to write `match`/`case` statements. Translating to Mint it'd be as simples as writing:
```
case(maybeSomething) {
  Just(something) => doSomethingWith(something)
  Nothing => doSomethingElse()
}

```
Notice `Maybe::Nothing` turned into a much simple `Nothing`. In Elm, if you write it like that and it finds more than one definition of `Nothing` it'll complain at compile-time and tell you to write the scope properly. Returning to the Blog example I've laid out above, that'd mean you'd only need to write the full scope when calling a Thumbnail, and you'd simply write `Category` and let the compiler figure out it's full/scoped name is actually `Post.Category`. I think if we want to avoid repetition in Mint we should probably go this way, instead of considering un-scoping methods on a case-by-case-basis, which will not only greatly increase the chances of something bad happening, but also results in re-enacting this discussing for every case, which compounds into a lot of time. The new `|.` operator was not a bad suggestion, but I think it's still not as good as a way to access unscoped methods/classes/enums/etc., and actually there's already something like that in Mint, when connecting Stores, as seen in https://www.mint-lang.com/guide/reference/components/connecting-stores (i.e., the line `connect Counter exposing { count as myCount }`), it could be solved by something like `from Time unscope {seconds,minutes,hours,days,weeks,months,years}`, or like Rust does it, `use Time::{seconds,minutes,hours,days,weeks,months,years}`, which could be even easier if those were in a separate scope and we had the option of using `*` to match everything in that scope, which could turn it into something as short as `use Time::Span::{*}` or, using the first notation that occurred to me, `from Time::Span unscope {*}`. That's how I'd approach this, which would fix all the cases we've been talking about where the pipes stay in the same module, but beyond that it'd solve the situation where you wanna pipe stuff to another module, then some other module, then back etc., which wouldn't be improved at all with the `|.` operator. This is not to say it's not a good idea to have the proposed `|.` operator, maybe it's a good ideal to ALSO have it, I just think that, first, a way to unscope methods has more applications and deserves a higher priority, and second, my understanding of the impacts, uses, advantages and possible issues with having the `|.` operator is insufficient. I don't know what languages have something like that, how people use it, basically what are the implications of having something like that in a language. I do have experience with Elm and Rust which provide ways to omit the scopes, and I can say with confidence that it makes things a lot easier.

If you do have examples, experience, reasons, information or whatever that leads you to believe something like `|.` is still something good to have, I encourage you to open an issue about that, but in the event it was just some wild brainstormy idea about a way to make the code less repetitive, we'll probably agree it's safer to use other languages as a reference before deciding what is a better fit for Mint. Personally, I'd go for a variation of `connect Store exposing {methods}`, only instead of a Store it would apply to a Module (Or whatever else this may apply to, though I can't recall anything other than Modules).

Either way, I do think this deserves an issue, and if you don't open one regarding either the `|.` operator or a way to reduce repetition, I'll create one after polishing this idea a bit and checking how I might go about implementing it.

Reply:
No need to have either a new operator `|.` or a new language feature for importing thinks in scope.

In this branch https://github.com/mint-lang/mint/tree/syntax-rewamp-3 there are features that tackle these issues:

- Short enum options `:Day(1)` - it will raise an error if it's ambiguous
- OO like function calls  `["a"].first().withDefault("") == "a"` - it will raise an error if it's ambiguous

Like I said in the other reply, soon I'll create a PR or discussion about these changes, but for now let's keep the discussion here about the `Time` module.